sql/sql_parse.cc

```diff
@@ -6362,6 +6362,72 @@ mysql_execute_command(THD *thd)
   DBUG_RETURN(res || thd->is_error());
 }
 
+struct String_pair
+{
+  String_pair(const char *f, const char *s) : first(f), second(s) {}
+  const char *first;
+  const char *second;
+};
+
+static bool get_table_fields(THD *thd, LEX_STRING db, LEX_STRING table_name,
+                             Dynamic_array<const char *> &fields)
+{
+  TABLE_LIST tl;
+  tl.init_one_table(LEX_STRING_WITH_LEN(db), LEX_STRING_WITH_LEN(table_name),
+                    table_name.str, TL_READ);
+  Open_tables_backup open_tables_backup;
+  thd->reset_n_backup_open_tables_state(&open_tables_backup);
+  bool can_deadlock= thd->mdl_context.has_locks();
+  bool error= open_tables_only_view_structure(thd, &tl, can_deadlock);
+  if (!error)
+  {
+    fields.clear();
+    TABLE_SHARE *s= tl.table->s;
+    for (size_t i= 0; i < s->fields; i++)
+    {
+      if (fields.append(s->field[i]->field_name))
+        return true;
+    }
+  }
+  close_thread_tables(thd);
+  MDL_savepoint &savepoint= open_tables_backup.mdl_system_tables_svp;
+  thd->mdl_context.rollback_to_savepoint(savepoint);
+  thd->restore_backup_open_tables_state(&open_tables_backup);
+
+  return error;
+}
+
+static bool get_fields_mapping(THD *thd, LEX_STRING db, LEX_STRING current_name,
+                               LEX_STRING archive_name,
+                               Dynamic_array<String_pair> &map)
+{
+  Dynamic_array<const char *> current_fields;
+  Dynamic_array<const char *> archive_fields;
+
+  if (get_table_fields(thd, db, current_name, current_fields))
+    return true;
+  if (get_table_fields(thd, db, archive_name, archive_fields))
+    return true;
```

Comment:
So much needless data shuffle: put fields in 2 arrays, converge 2 arrays into 1, cycle that array. And that's all inside single call. Why do you need all that?

Reply:
This temporary code will be removed (except for `map_field()`) once `CMMD` is done.

Reply:
This needless code should be removed now. You already have simple index mapping (see task description).